torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,352 @@
+                            buf.set_(tensor)
+
+    def _register_grad_hooks(self):
+        self._grad_accs = []  # need to keep them in scope
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                if p.requires_grad:
+                    p_tmp = p.expand_as(p)
+                    grad_acc = p_tmp.grad_fn.next_functions[0][0]
+                    grad_acc.register_hook(self._make_param_hook(p, device_idx))
+                    self._grad_accs.append(grad_acc)
+
+    def _make_param_hook(self, param, device_idx):
+        bucket_idx = self.bucket_map[param]
+
+        def distributed_data_parallel_hook(*unused):
+            if param.grad.requires_grad:
+                raise RuntimeError("DistributedDataParallelC10d only works "
+                                   "with gradients that don't require grad")
+            bucket = self.buckets[bucket_idx][device_idx]
+            bucket.append(param.grad.data)
+
+            # We can flush these and save memory for replicas
+            if device_idx > 0:
+                param.grad = None
+                param.data.set_()
+
+            # Current device's bucket is full
+            if len(bucket) == self.bucket_sizes[bucket_idx]:
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
+                    return
+
+                if bucket_idx not in self.buckets_ready:
+                    self.buckets_ready.add(bucket_idx)
+                    if len(self.buckets_to_reduce) > 0 and self.buckets_to_reduce[0] == bucket_idx:
+                        # Reduce the bucket
+                        self._queue_reduction(bucket_idx)
+                        self.buckets_to_reduce.pop(0)
+
+                    # If all buckets are ready
+                    if len(self.buckets_ready) == len(self.bucket_order):
+                        # In case there are unreduced buckets, reduce them all
+                        if len(self.buckets_to_reduce) > 0:
+                            for b_idx in self.buckets_to_reduce:
+                                self._queue_reduction(b_idx)
+                        # A final sync for all the reduction works
+                        self._sync_reduction_works()
+
+        return distributed_data_parallel_hook
+
+    def _queue_reduction(self, bucket_idx):
+        grads_batch = self.buckets[bucket_idx]
+        grads_batch_coalesced = []
+
+        # coalesce the bucket
+        for dev_id, dev_grads_batch in zip(self.device_ids, grads_batch):
+            with torch.cuda.device(dev_id):
+                dev_grads_batch_coalesced = _flatten_dense_tensors(dev_grads_batch)
+                grads_batch_coalesced.append(dev_grads_batch_coalesced)
+
+        # reduce to the first GPU in self.device_ids
+        if len(self.device_ids) > 1:
+            nccl.reduce(grads_batch_coalesced, root=0, streams=self.default_streams)
```

Comment:
Isn't it faster to let NCCL handle this part?

Reply:
yeah, as you can see in our discussion in the c10d thread, I am not able to get overlap if NCCL op is operated on more than one GPU from a single thread. cc @csarofeen @ngimel